AY-979 Finalize updating Resolve to new publisher

robin-ynput opened the pull request:
## Changelog Description

Finishes up with Resolve switch to new Publisher.
Finishes up the features that were already implemented in https://github.com/ynput/ayon-core/pull/157

- [x] - @BigRoy suggested warning message if there is not anything to be used for creation
- [x] - @BigRoy flagged that when clicking Create for Create Publishable Clip but there's no valid clips in timeline nothing happens
- [x] - @BigRoy flagged that previous creators was creating instances with invalid folder path(s) 
- [x] - @BigRoy flagged issues when deleting instances in the creator
- [x] - Spliting product types into:
        	- **primary**: `shot` - instance with hierarchy, future context data and attributes placeholder
            - **secondary**: `plate`, `audio`, `review` - instance with media
- [x] - Hierarchy and context data should be linked from Shot instance for all secondary product types Traypublisher (TP)
- [x] - Each secondary product type which will be created from a clip should have own creator class - like in TP
- [x] - When possible OTIO export should be done via native function rather then our implementation
- [x] - OTIO support for `opentimelineio >= 0.16.0`
- [x] - Ensure that audio cannot be exported from non-audio timeline(s)
- [x] - Test out with Resolve 19
- [x] - Help description in "?" button
- [x] - Fix review product not published when review_track contains a space
- [x] - Update client/README documentation


## Testing notes:

In Resolve:
1. Create a timeline with clips associated to a media
2. (Optional) Make a selection coloring some clips in *Chocolate* color
3. From `Workspace -> Script -> AYON Menu`
4. Create and publish product(s) of type shot, audio, plate and review from there

Tested locally on Windows with Resolve `19.0, 18.5` and OpenTimelineIO `0.17.0, 0.16.0`.

robin-ynput commented on client/ayon_resolve/plugins/publish/precollect_plates.py:
```diff
@@ -0,0 +1,20 @@
+        # Temporary disable no-representation failure.
+        # TODO not sure what should happen for the plate.
+        instance.data["folderPath"] = instance.data.pop("hierarchy_path")
+        instance.data["integrate"] = False 
```
Plate generation did not work for me:
![plate_norepresentation](https://github.com/user-attachments/assets/015d5d11-a018-4f9d-aa08-6d42b306e8c5)

For now I've disabled its integration in Resolve as I believe this is something common between Hiero, Resolve and Flame that would need to be addressed in `ayon_core`. Please let me know if I'm wrong.

jakubjezek001 replied:
To integrate clip resources effectively in the plugin `ayon-core\client\ayon_core\plugins\publish\collect_otio_subset_resources.py`, we need to include the `clip` family in the `instance.data["families"]`. This will ensure that the conversion process accounts for the clip resources as representations that can be integrated smoothly.

robin-ynput replied:
addressed !

robin-ynput commented on client/ayon_resolve/plugins/publish/precollect_shots.py:
```diff
@@ -0,0 +1,86 @@
+        https://github.com/ynput/ayon-traypublisher/blob/develop/client/ayon_traypublisher/plugins/publish/collect_shot_instances.py#L188
+        """
+        instance.data["folderPath"] = instance.data.pop("hierarchy_path")
+        instance.data["integrate"] = False  # no representation for shot
```
This is more of a work-around for now, but probably something worth to discuss with @jakubjezek001 and  @iLLiCiTiT.
* `folderPath` as to be set to an existing folder for the Creator to detect the instance has valid context
* `folderPath` has to be the hierarchy of the shot to create for "Collect Hierarchy" plugin to properly [compute the final path](  https://github.com/ynput/ayon-core/blob/6a07de6eb904c139f6d346fd6f2a7d5042274c71/client/ayon_core/plugins/publish/collect_hierarchy.py#L65)

As a work-around this code:
1. Create a instance with `folderPath` containing the selected parent context folder at Creation to produce a valid instance
2. Immediately replace `folderPath` with final hierarchy to create in early collect plugins

jakubjezek001 replied:
Once `newHierarchyIntegration` will be included you can fill `folderPath` value with Ayon marker value `folder_path`

robin-ynput replied:
I guess there is something I'm doing wrong then.
I have set `newHierarchyIntegration = True` to each of the created instances.

Here is what I get if I set `folderPath` to Ayon marker value `folder_path` when creating the instance:
![invalid_context](https://github.com/user-attachments/assets/bd86549a-0e93-4d37-94a5-a175eede5800)

This is what I get with my hack
(set `folderPath` to parent context when creating it then adjust it at precollect time):
![valid_context](https://github.com/user-attachments/assets/eabad5a7-6170-4901-9606-74369d76f66f)

jakubjezek001 replied:
Yeah you are quite right, the creating context is not supporting this attribute. So the hack is actually the same workaround as it is done in TP. @iLLiCiTiT question is if we could improve the workflow now so we do not need to hack this?

iLLiCiTiT replied:
Not sure when this PR should be merged? Not sure if I could get to it in next 2 weeks.

robin-ynput replied:
This PR is adjusted to report changes from https://github.com/ynput/ayon-core/pull/894/files

robin-ynput commented on client/ayon_resolve/plugins/create/create_shot_clip.py:
```diff
@@ -1,23 +1,136 @@
+#    create_allow_context_change = False  
+# TODO: explain consequence on folderPath
+# https://github.com/ynput/ayon-core/blob/6a07de6eb904c139f6d346fd6f2a7d5042274c71/client/ayon_core/tools/publisher/widgets/create_widget.py#L732
+
```
If `create_allow_context_change = False` then the create instance_data does not contain the selected `folderPath`.
See: https://github.com/ynput/ayon-core/blob/6a07de6eb904c139f6d346fd6f2a7d5042274c71/client/ayon_core/tools/publisher/widgets/create_widget.py#L732

This looks like a bug to me, would you like me to create an issue for that ?
Is there a better work-around than simply making it editable ?

jakubjezek001 replied:
Good question for @iLLiCiTiT - related to my comment here https://github.com/ynput/ayon-resolve/pull/15#discussion_r1743207592.

iLLiCiTiT replied:
If it is set to `False`, then it is job of create plugin to fill it.

robin-ynput replied:
Thanks for the clarification.

robin-ynput commented on client/ayon_resolve/plugins/create/create_workfile.py:
```diff
@@ -17,8 +22,61 @@ class CreateWorkfile(AutoCreator):
+            constants.ayon_marker_name,
+            note,
+            constants.ayon_marker_duration
+        )
```
@jakubjezek001 Unfortunately it does not look like Resolve allow to register custom metadata at Timeline/Track level (plus Metadata gets totally ignored by the native OTIO exporter right now).

Best solution I could think of was to store the serialized workfile instance along timeline resolution metadata as a marker on the first frame of the timeline itself. 
![image](https://github.com/user-attachments/assets/363754e6-147d-46ba-89e3-7984bda1d1b8)
 
 Is that a good idea ? Do you have a better idea/recommendation ?

BigRoy replied:
In all hosts as far as I know the "Workfile" represents the full scene file of the host application. In this case it looks like we're considering only one timeline, and that is the _active_ timeline. Which may be odd behavior? Do we need to store this data PER timeline? Or do we need it once in the full project. If the latter, then...

There doesn't seem to be any API to set `Project Notes` otherwise that could've been handy to put a JSON string.
What if we make a dedicated "Disabled" timeline that we solely use to store the metadata.

![image](https://github.com/user-attachments/assets/786fc097-d6d6-4eee-ba19-e0af3f0bd219)

jakubjezek001 replied:
I'm with @BigRoy on this. Instead of the current approach, we could create a sidecar .JSON file in the work folder alongside the .DRP files. This would serve as our go-to for reading and storing metadata. As for clip-level data, we can keep using markers like we do now.

BigRoy replied:
Sidecar files have a lot of issues in management - I rarely feel that it's worth it going that route.

It would really be much nicer if we can find a _hacky_ way to store it _within_ the resolve project.

For what it's worth - it may be worth it for you all to make an account on Blackmagic Design forum and just respond to [this topic](https://forum.blackmagicdesign.com/viewtopic.php?f=21&t=189685&hilit=python+database#p987221) to get some visibility on that by Resolve devs, etc. plus maybe the topic already has some other hints. (Ignore the guy who created that topic!)

And look at [this hack mentioned there](https://forum.blackmagicdesign.com/viewtopic.php?f=21&t=189685&hilit=python+database#p991541). Could that suit our needs?

jakubjezek001 replied:
> And look at [this hack mentioned there](https://forum.blackmagicdesign.com/viewtopic.php?f=21&t=189685&hilit=python+database#p991541). Could that suit our needs?

Yeah! This is definitely the better way. It actually got the attention in API doc but since there were not list of available  keys and value type. Thanks for the tip @BigRoy

jakubjezek001 commented on client/ayon_resolve/plugins/create/create_shot_clip.py:
```diff
@@ -48,42 +161,22 @@ def header_label(text):
-                default=presets.get("countSteps", 10),
+                label=header_label("Additional Export(s)")
             ),
+            BoolDef("export_plate",
+                    label="Plate",
+                    tooltip="Export Plate output(s)",
+                    default=False),
+            BoolDef("export_review",
+                    label="Review",
+                    tooltip="Export Review output(s)",
+                    default=False),
+            BoolDef("export_audio",
+                    label="Audio",
+                    tooltip="Export Audio output(s)",
+                    default=False),
```
I suggest we remove the section related to Traypublisher editorial workflow. In Hiero/Resolve/Flame, we can address this by using a combination of `productType`, `reviewTrack`, and `audio` fields. Currently, we need to decide how to handle the `take` type in the `productType` enum field, as it's not yet implemented.

robin-ynput replied:
@jakubjezek001 sounds good to me ! Would you mind to confirm the expected UX please ?
Just to be sure I got it right. (For now I replicated the exact same behavior then current publisher.)

![image](https://github.com/user-attachments/assets/a13d7f4c-3d20-4139-a984-fd1ec9110928)

* I'm not sure what **Use Review Track** is needed for actually and couldn't really figure out from the code
* **Renaming** seems to be applied only to the shot name but not the resulting plate product(s) name which remain `{product_type}{variant}` (e.g. plate_VideoTrack)
* When creating a timeline from movie clips, **Include Audio** did not seem to have any effect. 
_So I've implemented a new plugin to create "audio" product from the clip source range (no handles) but that might not be what's needed._

jakubjezek001 replied:
Here are some notes:

![image](https://github.com/user-attachments/assets/9d1880a4-fbb3-46d4-8bbb-f24818dd6998)

> I'm not sure what Use Review Track is needed for actually and couldn't really figure out from the code

From the image above you can see that additional track could be added to the timeline and that can be then pointed to the **Use Review Track** field. 

> Renaming feature seems to be applied only to the shot name but not the resulting plate product(s) name which remain {product_type}{variant} (e.g. plate_VideoTrack), should we re-label as "Rename Shot" ?

Yes, that's correct for now. It works just as you described. The product name is created by combining the Product Variant field and Product Type like this: `{product_type}{Product_variant}`. We're not using product name presets at the moment because we're still deciding how to handle tasks in the future. We might need to introduce a specific task for plate publishing, as we've seen several client requests for this in the past.

> When creating a timeline from movie clips, Include Audio did not seem to have any effect. Meaning the extracted plate got audio whatever it's checked or not.
So I've implemented a new plugin to create "audio" product from the clip source range (no handles) but that might not be what's needed.

To be honest I am not sure if it ever worked. This had been migrated from Hiero but since noone was ever reporting it it might just slip our attantion. Correct behaviour should be following:

1. Include audio toggled on
2. Additional `audioMain` product instances per clip should be created
3. Publishing process should be coverting whole timeline (included audio tracks) into OTIO and then other Core plugin should use FFMPEG and covert flattened tracks into single interemediate file which is then trimmed by the original clip timing. https://github.com/ynput/ayon-core/blob/develop/client/ayon_core/plugins/publish/extract_otio_audio_tracks.py

So I am assuming that main issue here was that we were ignoring audio tracks during our custom OTIO exporter.

robin-ynput replied:
Awesome, thanks for confirming.

jakubjezek001 commented on client/ayon_resolve/plugins/publish/precollect_plates.py:
```diff
@@ -0,0 +1,20 @@
+        # Temporary disable no-representation failure.
+        # TODO not sure what should happen for the plate.
+        instance.data["folderPath"] = instance.data.pop("hierarchy_path")
+        instance.data["integrate"] = False 
```
```suggestion

```

BigRoy replied:
@jakubjezek001 Can you explain this suggestion?

jakubjezek001 replied:
We also have to include following instance data key for every instance which is having nonexistent hierarchy at the moment of publishing. 

`instance.data["newHierarchyIntegration"] = True`

robin-ynput replied:
Yep I had the "newHierarchyIntegration" set but I can confirm adding the "clip" family to plate now gives it a representation.

robin-ynput commented on client/ayon_resolve/plugins/publish/precollect_review.py:
sure thing

BigRoy reviewed (commented):
Testing with Resolve 19.0B build 25 (beta release) on Windows 10.

### Publisher UI errors on empty project

Opening the publisher UI in an empty project gives me:

![image](https://github.com/user-attachments/assets/89216d37-a9bb-4090-b0e3-a6a9dbd49f97)

```
Creator: Workfile (io.ayon.creators.resolve.workfile)

Error: 'NoneType' object has no attribute 'GetMarkers'

Traceback (most recent call last):
  File "E:\dev\ayon-core\client\ayon_core\pipeline\create\context.py", line 1009, in reset_instances
    creator.collect_instances()
  File "E:\dev\ayon-resolve\client\ayon_resolve\plugins\create\create_workfile.py", line 120, in collect_instances
    data = self._loads_data_from_marker()
  File "E:\dev\ayon-resolve\client\ayon_resolve\plugins\create\create_workfile.py", line 72, in _loads_data_from_marker
    _, marker_info = self._get_timeline_marker()
  File "E:\dev\ayon-resolve\client\ayon_resolve\plugins\create\create_workfile.py", line 61, in _get_timeline_marker
    for idx, marker_info in timeline.GetMarkers().items():
AttributeError: 'NoneType' object has no attribute 'GetMarkers'
```

---

### Do not store data into the workfile on just opening publisher UI

Opening the creator or publisher UI _directly_ stores metadata into the scene. This should instead happen on the first `save` of the publisher UI. As such, I'd say data should persist into the file preferably only on `update_instances()` if we can instead of directly creating on `collect_instances()`.

That way just opening the UI doesn't suddenly 'pollute' new stuff into your workfile.

---

### Allow more clip color filters using enum 

(could be follow up PR, but may be trivial to implement)

I wonder if it would make sense to change this "Create" toggle from 'Use clips of Chocolate color' to an ENUM that instead is:

- Filter to clip colors: `["All clips", "Red", "Blue", "White"]`

So that basically we're not enforced to that chocolate color.

---

### Multiple publish instances

There are likely reasons for it, but I'm personally a bit confused about the double instances visible in the UI. With one clip I get:

![image](https://github.com/user-attachments/assets/13eebc82-1c94-4407-8b45-d1e142b39e05)

Why do we need a separate "Editorial Plate" and "Editorial Shot" instance? Can't those be one publish instance? Also, what happens if I disable in the UI only editorial shot for example?

---

### Older OTIO

I got this error:
```
Traceback (most recent call last):
  File "E:\dev\ayon-resolve\client\ayon_resolve\otio\utils.py", line 106, in get_marker_from_clip_index
    all_clips = otio_timeline.find_clips()
AttributeError: 'opentimelineio._otio.Timeline' object has no attribute 'find_clips'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\User\AppData\Local\Ynput\AYON\dependency_packages\ayon_2406251801_windows.zip\dependencies\pyblish\plugin.py", line 528, in __explicit_process
    runner(*args)
  File "E:\dev\ayon-resolve\client\ayon_resolve\plugins\publish\precollect_plates.py", line 26, in process
    otio_clip, _ = utils.get_marker_from_clip_index(
  File "E:\dev\ayon-resolve\client\ayon_resolve\otio\utils.py", line 108, in get_marker_from_clip_index
    all_clips = otio_timeline.each_clips()
AttributeError: 'opentimelineio._otio.Timeline' object has no attribute 'each_clips'
```
But I suspect that's due to using an older OTIO library. I believe historically there were reasons to stick with the older version in the Resolve integrations because it didn't work with the newer ones.

We should EITHER make it super clear that with this `ayon-resolve` release you need a newer OTIO. Or make this backwards compatible and work with the older?

I was running with `OpenTimelineIO 0.15.0`
When updating to `0.17.0` the error was gone.

---

Not sure if relevant, but I was also getting this warning in publisher UI:

```
Could not find root path for remapping "ynts_char_hero_animation_v001". This may cause issues on farm.
```

![image](https://github.com/user-attachments/assets/01f64c9e-f334-4d3e-adf7-b7243a164a17)

BigRoy commented on client/ayon_resolve/plugins/create/create_shot_clip.py:
```diff
@@ -223,7 +339,25 @@ def create(self, subset_name, instance_data, pre_create_data):
+            if pre_create_data.get("use_selection", False):
+                raise CreatorError(
+                    "No Chocolate-colored clips found from "
+                    "timeline.\n\n Try changing clip(s) color "
```
```suggestion
                    "timeline.\n\nTry changing clip(s) color "
```

Fix space on second line.
![image](https://github.com/user-attachments/assets/b0888b06-a608-4643-a956-318254dc72a7)

After:
![image](https://github.com/user-attachments/assets/90625036-649e-4f04-9ff8-0242d57b9dcb)

BigRoy commented on client/ayon_resolve/plugins/create/create_shot_clip.py:
```diff
@@ -1,23 +1,142 @@
+            track_item = instance.transient_data["track_item"]
+            tag_data = lib.get_timeline_item_ayon_tag(track_item)
+            instances_data = tag_data.get(_CONTENT_ID, {})
+            _ = instances_data.pop(self.identifier, None)
```
```suggestion
            instances_data.pop(self.identifier, None)
```

BigRoy commented on client/ayon_resolve/plugins/create/create_shot_clip.py:
```diff
@@ -240,95 +374,105 @@ def create(self, subset_name, instance_data, pre_create_data):
         # create media bin for compound clips (trackItems)
         media_pool_folder = create_bin(self.timeline.GetName())
 
+        # detecte enabled creators for review, plate and audio
```
```suggestion
        # detect enabled creators for review, plate and audio
```

BigRoy commented on client/ayon_resolve/plugins/create/create_shot_clip.py:
```diff
@@ -240,95 +374,105 @@ def create(self, subset_name, instance_data, pre_create_data):
-            self.log.info(f"Storing data: {data}")
-
-            lib.imprint(track_item, data)
+        """Never called, update is handled via _ResolveInstanceCreator."""
```
Handled by the method definition on the parent class? In what way? Would that mean you just wouldn't re-implement it here instead? or?

robin-ynput replied:
No, sorry my explanation is probably misleading. 
Basically this `CreateShotClip` creator always delegates instance creation to "sub-creators" which are instances of `_ResolveInstanceCreator` but never hold any instances itself.
Which means that while we need `update_instances`, `remove_instances` and such defined to keep it valid creator, those will never be called from the outside. 

Operations will be handled by the "sub" creators which actually generated the instance:
* update/remove of `shot` will be made through  `ResolveShotInstanceCreator(_ResolveInstanceCreator)`
* update/remove of `plate` will be made through  `EditorialPlateInstanceCreator(_ResolveInstanceCreator)`

Hopefully this clarify a bit, let me know how I could rephase those docstring maybe.

jakubjezek001 commented:
> Allow more clip color filters using enum

Yeah that would be great but lets add it as enhancement issue for now please.

jakubjezek001 commented:
> Could not find root path for remapping "ynts_char_hero_animation_v001". This may cause issues on farm.

This should not be problem, since Resolve for now is not processed over farm. But this could be fixed by adding source path to the work .DRP file.

robin-ynput commented:
> > Allow more clip color filters using enum
> 
> Yeah that would be great but lets add it as enhancement issue for now please.

https://github.com/ynput/ayon-resolve/issues/17

iLLiCiTiT commented on client/ayon_resolve/api/constants.py:
```diff
@@ -2,6 +2,7 @@
 rename_index = 0
 rename_add = 0
 
+selected_clip_color = "Chocolate"
```
I know that you've added only this constant (and you followed what was here), but it would nice to use upper case variable names, for constants.

jakubjezek001 replied:
Agreed. And you can also change to uppercase the rest, please.

robin-ynput replied:
Sure thing.

iLLiCiTiT commented on client/ayon_resolve/api/lib.py:
```diff
@@ -958,8 +991,8 @@ def get_otio_clip_instance_data(otio_timeline, timeline_item_data):
 
-def get_otio_temp_dir(project_name, anatomy=None, timeline=None) -> str:
-    """Get otio temporary directory.
+def _get_otio_temp_file(project_name=None, anatomy=None, timeline=None) -> str:
```
Project name should be requirement.

robin-ynput replied:
Updated !

iLLiCiTiT commented on client/ayon_resolve/api/plugin.py:
```diff
@@ -7,7 +8,8 @@
 from ayon_core.pipeline import (
     LoaderPlugin,
-    Creator as NewCreator
+    Creator as NewCreator,
```
Can this be changed to be only `Creator`?

robin-ynput replied:
Done !

iLLiCiTiT commented on client/ayon_resolve/api/plugin.py:
```diff
@@ -454,6 +456,13 @@ def convert(self):
+        # AYON unique identifier
+        folder_path = "/{}/{}".format(
+            self.tag_data["hierarchy"],
+            self.tag_data["asset"],
```
Not sure how much work it would be, but this should not be `"asset"` (openpype term) but `"folder_name"`.

jakubjezek001 replied:
Yes that is residual OP key name, but we have to keep this backward compatible when it comes to loading already created markers from previouse addon versions. Some clients might need transition time for their already created project markers. Lets clearly comment those backward compatibility changes for future references. 

Here is how some comment looks `# Backwards compatible (Deprecated since 24/06/06)` but @BigRoy might have some better idea on how this should be done. 

BTW, there is a lot of backward compatibility code coments which could be already removed.

robin-ynput replied:
Oh ! As I flagged in the PR header I didn't put too much attention in making the marker data backward compatible. 
But if that's a thing I'll definitely ensure it is!

robin-ynput replied:
Just pushed some new commits to ensure these changes remains compatible and can load markers generated from `OpenPypeData`.

iLLiCiTiT commented on client/ayon_resolve/api/plugin.py:
```diff
@@ -746,3 +745,13 @@ def _store_new_instance(self, new_instance):
         self._add_instance_to_context(new_instance)
+
+
+def get_representation_files(representation):
```
Where is this used?

jakubjezek001 replied:
https://github.com/ynput/ayon-resolve/blob/feature/AY-979_finish_Resolve_new_publisher/client/ayon_resolve/plugins/load/load_clip.py#L43-L44

robin-ynput replied:
Yep this is not new code, I realized this was missing from parent PR when testing the load features.

iLLiCiTiT commented on client/ayon_resolve/api/plugin.py:
```diff
@@ -454,6 +456,13 @@ def convert(self):
+            self.tag_data["hierarchy"],
+            self.tag_data["asset"],
+        )
+        self.tag_data["folder_path"] = folder_path
```
Does this have to be set on created instance? This can be collected during publishing.

robin-ynput replied:
just to flag if we rename it "folderPath" right now, it creates invalid-context instances in the publisher.

iLLiCiTiT replied:
I think he cannot rename it to `folderPath` because publishe UI would complain that it does not exist.

robin-ynput replied:
OK, so I think I've this:

* I temporarily stored what used to be set in `folder_path` into a temporary `target_folder_path` entry, so I do not overwrite `folderPath` which contains the value needed to keep the instance valid but store it in the creator_attributes instead. So `target_folder_path` will not be part of the instance.
* Then, as per @jakubjezek001 recommendation, I've implemented the replication via `editorialSharedData` into the shot collector so that `folderPath` is properly set for each product upon collection time (same logic than traypublisher).

![image](https://github.com/user-attachments/assets/a53c5966-0643-473f-a53f-b2a22337aab8)

jakubjezek001 commented on client/ayon_resolve/plugins/publish/collect_current_project.py:
```diff
@@ -1,30 +1,29 @@
-        otio_timeline = davinci_export.create_otio_timeline(resolve_project)
+        video_tracks = api.get_video_track_names()
+        otio_timeline = api.export_timeline_otio(timeline)
+        current_file = resolve_project.GetName()
```
```suggestion
        host = registered_host()
        current_file = host.get_current_workfile()
```

And also add: 
- `from ayon_core.pipeline import registered_host`

jakubjezek001 replied:
this should fix the issue from here: https://github.com/ynput/ayon-resolve/pull/15#issuecomment-2328493576

robin-ynput replied:
this is updated !

jakubjezek001 commented:
Lets also include into this PR following improvement:

Solution form here https://github.com/ynput/ayon-resolve/pull/15/files#r1745021400 is also applicable at https://github.com/ynput/ayon-resolve/blob/feature/AY-979_finish_Resolve_new_publisher/client/ayon_resolve/plugins/publish/extract_workfile.py lines 27 to 30.

robin-ynput commented on client/ayon_resolve/plugins/create/create_shot_clip.py:
```diff
@@ -240,95 +375,105 @@ def create(self, subset_name, instance_data, pre_create_data):
-            self.log.info(f"Storing data: {data}")
-
-            lib.imprint(track_item, data)
+        """Never called, update is handled via _ResolveInstanceCreator."""
```
You got it

iLLiCiTiT commented on client/ayon_resolve/api/plugin.py:
```diff
@@ -746,3 +745,13 @@ def _store_new_instance(self, new_instance):
         self._add_instance_to_context(new_instance)
+
+
+def get_representation_files(representation):
+    anatomy = Anatomy()
```
```suggestion
def get_representation_files(project_name, representation):
    anatomy = Anatomy(project_name)
```

robin-ynput replied:
This is adjusted, thanks for the feedback.

iLLiCiTiT commented on client/ayon_resolve/api/plugin.py:
```diff
@@ -746,3 +745,13 @@ def _store_new_instance(self, new_instance):
+
+def get_representation_files(representation):
+    anatomy = Anatomy()
+    files = []
+    for file_data in representation["files"]:
+        path = anatomy.fill_root(file_data["path"])
+        files.append(path)
+    return files
```
```suggestion
    return [
        anatomy.fill_root(file_data["path"])
        for file_data in representation["files"]
    ]
```

robin-ynput replied:
reported !

BigRoy commented on client/ayon_resolve/api/constants.py:
```diff
@@ -1,19 +1,19 @@
 # Ayon sequential rename variables
-rename_index = 0
-rename_add = 0
+#rename_index = 0
+#rename_add = 0
```
Commented lines should start with a space after the `#` according to PEP style guide... (yes, very nitpicky; sorry!)
```suggestion
# rename_index = 0
# rename_add = 0
```


However, aside of that. Is there a reason these are now commented but still left in? Should they be removed?

BigRoy commented on client/ayon_resolve/api/lib.py:
```diff
@@ -906,6 +908,32 @@ def _convert_resolve_list_type(resolve_list):
+        width = height = None
+
+    try:
+        pixel_aspect = int(clip_properties["PAR"]) # Pixel Aspect Resolution
```
PEP 8: E261 at least two spaces before inline comment
```suggestion
        pixel_aspect = int(clip_properties["PAR"])  # Pixel Aspect Resolution
```

BigRoy replied:
Also, what's the value of "PAR"? Because, isn't pixel aspect ratio usually a float?

BigRoy commented on client/ayon_resolve/api/lib.py:
```diff
@@ -906,6 +908,32 @@ def _convert_resolve_list_type(resolve_list):
+
+    try:
+        pixel_aspect = int(clip_properties["PAR"]) # Pixel Aspect Resolution
+    except(KeyError, ValueError):
```
When would this error?

robin-ynput replied:
I looked into that a bit more, it's not quite clear what Resolve is doing. It seems to store PAR as a string.
![image](https://github.com/user-attachments/assets/66e54dc4-cab4-4377-9e5a-6f133aba6ccb)

So I've come ahead with this approach:
https://github.com/ynput/ayon-resolve/pull/15/files#diff-634f1d18d81de9d69a1b8517faa8142e794ee030651aee279a5982d806d20269R20
(btw this helped me a lot https://www.unravel.com.au/aspect-ratio-cheat-sheet)

I kept a default to `1.0` in case of a weird behavior from Resolve.

BigRoy commented on client/ayon_resolve/api/plugin.py:
```diff
@@ -148,7 +150,7 @@ def load(self, files):
         # create clip media
         media_pool_item = lib.create_media_pool_item(
-            files,
+            files[0],
```
So - what does this fix? It seems like an odd change to be related to "the new publisher" since it's a change for the loader?

robin-ynput replied:
You're right, this is totally not related to the "new publisher" thing. 
Testing stuff on my side, I realized "loading as a clip" feature was broken on my branch, and this was my quick fix to get it sorted.

Thanks to your comment I digged into it a bit more. And it comes from this:
https://github.com/ynput/ayon-resolve/commit/d3ba10caf1a3b3b3990bf00b39772824bad64a3a?diff=unified&w=0#diff-bae424ef949b94556f4ade49faccf3ad7d73f3c16135597bf20a081f9ebaa711R206
`lib.create_media_pool_item` has changed to accept a file and not a list anymore. So I've replicated the previous logic one step above so the overall load logic does not change.

BigRoy commented on client/ayon_resolve/api/plugin.py:
```diff
@@ -746,3 +745,20 @@ def _store_new_instance(self, new_instance):
+        representation (dict): The representation to inspect.
+
+    Returns:
+        list. The files associated to the representation.
```
```suggestion
        list: The files associated to the representation.
```

BigRoy commented on client/ayon_resolve/plugins/create/create_shot_clip.py:
```diff
@@ -1,23 +1,143 @@
+_CONTENT_ID = "resolve_sub_products"
+
+
+class _ResolveInstanceCreator(plugin.HiddenResolvePublishCreator):
```
@iLLiCiTiT should this "base class" be taken out of this file so the Plug-in discovery system does not even find it and starts thinking it's a creator it needs to "process" in one way or another being just bloat to the plug-in registry at that point?

iLLiCiTiT replied:
It can be. But I have nothing agains keeping it here if this is the only file using it.

BigRoy commented on client/ayon_resolve/plugins/create/create_workfile.py:
```diff
@@ -63,8 +102,12 @@ def create(self, options=None):
+
+        Args:
+            update_list(List[UpdateData]): Gets list of tuples. Each item
+                contain changed instance and it's changes.
```
```suggestion
                contain changed instance and its changes.
```

BigRoy commented on client/ayon_resolve/plugins/create/create_workfile.py:
```diff
@@ -52,9 +82,18 @@ def collect_instances(self):
+            self.product_type, data["productName"], data, self)
         self._add_instance_to_context(current_instance)
 
     def create(self, options=None):
```
I vaguely recall @iLLiCiTiT you mentioning to me that there was _still_ a reason to implement `create` for `AutoCreator` instead of doing the 'creation' inside `collect_instances()` but I can't remember what it was. I just remember fighting you about it haha, but then subsiding because you had an explanation for it needing it.

iLLiCiTiT replied:
I remember there was a reason, now I don't remember (it's 2 years). I would say as simple as `collect_instances` should only collect existing instances (unless there is some backwards compatibility handling), `create` should create them and store the metadata to file.

robin-ynput replied:
No problem. This is adjusted accordingly.

BigRoy commented on client/ayon_resolve/plugins/publish/collect_plates.py:
```diff
@@ -0,0 +1,32 @@
+        Args:
+            instance (pyblish.Instance): The shot instance to update.
+        """
+        instance.data["folderPath"] = instance.data["folder_path"]
```
Did we decide on keeping this `folder_path` instead of e.g. `targetFolderPath`?

robin-ynput replied:
This code is no more relevant, hopefully this shed some light on this:
https://github.com/ynput/ayon-resolve/pull/15#discussion_r1746214679

BigRoy commented on client/ayon_resolve/api/lib.py:
```diff
@@ -906,6 +908,34 @@ def _convert_resolve_list_type(resolve_list):
+        clip_par = clip_properties["PAR"]  # Pixel Aspect Resolution
+        pixel_aspect = constants.PAR_VALUES[clip_par]
+
+    except(KeyError, ValueError): # Unknown or undetected PAR
```
```suggestion
    except (KeyError, ValueError):  # Unknown or undetected PAR
```

jakubjezek001 commented on client/ayon_resolve/plugins/create/create_shot_clip.py:
```diff
@@ -240,95 +382,213 @@ def create(self, subset_name, instance_data, pre_create_data):
+                        "parent_instance_id": parenting_data["instance_id"],
+                        "creator_attributes": {
+                            "folderPath": shot_folder_path,
+                            "parent_instance": parenting_data["label"],
```
Lets convert this into `TextDef` with nice label. But let's also make sure this field in non editable by adding input attribute `disabled=True`. For reference look into `ayon-core\client\ayon_core\lib\attribute_definitions.py`

jakubjezek001 replied:
Lets actually do the same for the linked child product instances:
![image](https://github.com/user-attachments/assets/fb71671a-cf38-448f-a58a-5f42f13916ef)

I really like how TP is having this done via shared clip base class and the TextDef is added to the class methon `def get_instance_attr_defs`. Here is some reference: [ayon_traypublisher/plugins/create/create_editorial.py#L90-L119](https://vscode.dev/github/ynput/ayon-traypublisher/blob/develop/client/ayon_traypublisher/plugins/create/create_editorial.py#L90-L119)

jakubjezek001 replied:
linking here the TP pr for above suggestions https://github.com/ynput/ayon-traypublisher/pull/12

robin-ynput replied:
Sounds good to me, thanks for the links. Here is what I obtain :

For Plate / Audio:
![image](https://github.com/user-attachments/assets/42d27f9b-356a-496b-a3d4-b99927ad0eb9)
Is the TP "Review" attribute also needed here ? again apologies but Im not entirely sure about this one.

For Shots:
![image](https://github.com/user-attachments/assets/39267df4-ca3e-4732-ae48-c63272aaf9ef)

I've left out the clipIn, clipDuration, clipOut.. read only attributes This way there is no need to query/compute all of this information but maybe that's wrong.. 
My thought was that unlike an EDL import you can visually assess your clips directly from the timeline. 
Is that all right for you ?

jakubjezek001 replied:
> I've left out the clipIn, clipDuration, clipOut.. read only attributes This way there is no need to query/compute all of this information but maybe that's wrong..

I believe a user may want to compare and review the data in AYON to ensure its accuracy. This could save us processing time, but it might also increase the time users spend on validation, and so do we during development. I suggest we add clip and frame attributes in a read-only format to streamline this process.

BigRoy replied:
> I've left out the clipIn, clipDuration, clipOut.. read only attributes This way there is no need to query/compute all of this information but maybe that's wrong..

If we just had "Instance" **actions** in the Publisher UI we could have a "Select clip" action. 🎉

jakubjezek001 replied:
> Is the TP "Review" attribute also needed here ?

Absolutely, that's a valid point. In Resolve or Hiero, we'll have to approach things differently compared to TP. TP has a single track where plates can be flagged for review. For Resolve and Hiero, we should introduce an editable dropdown field on the hero track clips. The dropdown options should come from the names of the timeline tracks. This value is initially set during the creation phase, but users may need to adjust it later if they forget.

Additionally, we'll need to implement a new toggle called "Hero Clip Enabled" (or something along those lines). This feature is crucial for managing vertically aligned clips and is another setting users might need to modify at a later stage.

robin-ynput replied:
Here is what I got for Shot:
![image](https://github.com/user-attachments/assets/29629309-b9ef-47ff-adc1-2623dc9d859a)

for plate:
![image](https://github.com/user-attachments/assets/0832c793-7e97-4d79-8da5-361d0af2fbcc)

(note that those toggle have currently *no affect* on the plate creation, I think I'd need a meeting or a call with you @jakubjezek001 if more work is expected for this PR about the review product)

and audio product:
![image](https://github.com/user-attachments/assets/f3196fe3-b8f1-44ab-95d2-0b4e56a40bc9)


Does this work for you guys ?

jakubjezek001 commented on client/ayon_resolve/plugins/publish/extract_workfile.py:
```diff
@@ -16,29 +18,22 @@ class ExtractWorkfile(publish.Extractor):
-        drp_file_path = os.path.normpath(
-            os.path.join(staging_dir, drp_file_name))
+        host = registered_host()        
+        drp_file_path = host.get_current_workfile()
```
We could also just get the already collected path from here https://github.com/ynput/ayon-resolve/pull/15/files#diff-d7c7c39f26e05a52a09cc2380f7dbe5c82a160e83ed1118b6fb7cca9925666c7R30

robin-ynput replied:
Good call, thanks for that.

jakubjezek001 reviewed (changes requested):
There are some comments and also the publishing had failed on missing `files` in plate related representation. Following is the publishing report, you can use 

[publish-report-240910-10-19.json](https://github.com/user-attachments/files/16941802/publish-report-240910-10-19.json)


`.\tools\manage.ps1 run publish-report-viewer --use-dev` and drop the json into it to see more details.

jakubjezek001 commented:
Hey @robin-ynput I had changed base target branch to Develop and closed the original #3 since all changes were already included here.

robin-ynput commented:
> There are some comments and also the publishing had failed on missing `files` in plate related representation. Following is the publishing report, you can use ayon-launcher in following way to view it.
> 
> [publish-report-240910-10-19.json](https://github.com/user-attachments/files/16941802/publish-report-240910-10-19.json)
> 
> `.\tools\manage.ps1 run publish-report-viewer --use-dev` and drop the json into it.

I've opened a separate issue for that, investigating it aside. I might be wrong but I think it actually comes from a wrong retimed computation by `ayon_core` from a valid OTIO data.

jakubjezek001 reviewed (changes requested):
I tested the latest changes with the latest dev on [this GitHub pull request](https://github.com/ynput/ayon-core/pull/897), but unfortunately, it did not go well. Here, I am including the publish report and the OTIO file it created for me.

[publish-report-240919-10-57.json](https://github.com/user-attachments/files/17056322/publish-report-240919-10-57.json)

[workfileConform_otio.zip](https://github.com/user-attachments/files/17056343/workfileConform_otio.zip)

I apologize if the review changes are more related to the [https://github.com/ynput/ayon-core/pull/897](https://github.com/ynput/ayon-core/pull/897). However, you might also find that they could be related to Resolve.

robin-ynput commented:
> I tested the latest changes with the latest dev on [this GitHub pull request](https://github.com/ynput/ayon-core/pull/897), but unfortunately, it did not go well. Here, I am including the publish report and the OTIO file it created for me.
> 
> [publish-report-240919-10-57.json](https://github.com/user-attachments/files/17056322/publish-report-240919-10-57.json)
> 
> [workfileConform_otio.zip](https://github.com/user-attachments/files/17056343/workfileConform_otio.zip)
> 
> I apologize if the review changes are more related to the [ynput/ayon-core#897](https://github.com/ynput/ayon-core/pull/897). However, you might also find that they could be related to Resolve.

 @jakubjezek001 just to follow up on this, the fix for this issue is in 
https://github.com/ynput/ayon-core/pull/916

iLLiCiTiT commented on client/ayon_resolve/README.markdown:
```diff
@@ -10,17 +10,17 @@
 -   make sure Resolve Fusion (Fusion Tab/menu/Fusion/Fusion Settings) is set to Python 3.6
     ![image](https://user-images.githubusercontent.com/40640033/102631545-280b0f00-414e-11eb-89fc-98ac268d209d.png)
--   Open OpenPype **Tray/Admin/Studio settings** > `applications/resolve/environment` and add Python3 path to `RESOLVE_PYTHON3_HOME` platform related.
+-   Open Ayon **Tray/Admin/Studio settings** > `applications/resolve/environment` and add Python3 path to `RESOLVE_PYTHON3_HOME` platform related.
```
It should be `AYON` instead of `Ayon`.

robin-ynput replied:
this is adjusted, thanks.

BigRoy commented on client/ayon_resolve/api/lib.py:
```diff
@@ -224,43 +223,58 @@ def create_bin(name: str,
-    # add all data in folder to media pool
-    media_pool_items = media_pool.ImportMedia(files)
+    dirname, file = os.path.split(fpath)
+    _name, ext = os.path.splitext(file)
 
-    return media_pool_items.pop() if media_pool_items else False
+    # add all data in folder to media-pool
+    media_pool_items = media_storage.AddItemListToMediaPool(
+        os.path.normpath(dirname))
```
Can we make this work without doing this? I'd really prefer it to NOT load ALL media from the folder to the media pool.

robin-ynput replied:
Yeah, I've restore current implementation. I think this changes were on initial draft PR but they do not seem relevant with this PR. I could confirm it all work successfully.

BigRoy commented on client/ayon_resolve/api/lib.py:
```diff
@@ -953,13 +1017,96 @@ def get_otio_clip_instance_data(otio_timeline, timeline_item_data):
+            timeline=timeline
+        )
+        otio_export.write_to_file(otio_timeline, filepath)
+
```
```suggestion
```

BigRoy commented on client/ayon_resolve/api/pipeline.py:
```diff
@@ -97,27 +99,33 @@ def get_workfile_extensions(self):
     def get_containers(self):
         return ls()
 
+    def get_context_data(self):
+        return {}
+
+    def update_context_data(self, data, changes):
+        pass
```
This is still lacking - and due to this any context option attributes in the publisher UI will not persist. Which isn't a big problem because apparently we do not have any context attributes exposed yet in the publisher UI, at least I didn't see any in resolve - but we might want to add todos here.

```suggestion
    def get_context_data(self):
        # TODO: implement to support persisting context attributes
        return {}

    def update_context_data(self, data, changes):
        # TODO: implement to support persisting context attributes
        pass
```

BigRoy replied:
It'd basically be an implementation similar to this https://github.com/ynput/ayon-resolve/blob/1b6e9aa353dc88d86d51ec3f41920e27ee0e441d/client/ayon_resolve/plugins/create/create_workfile.py#L24-L46 but then storing it in something that does not clash with that data.

robin-ynput replied:
This creator class is inherited by "exposed" AYON creators implemented in `plugins/create/create_shot_clip.py` file.
I believe the code you are looking for is there:
https://github.com/ynput/ayon-resolve/blob/feature/AY-979_finish_Resolve_new_publisher/client/ayon_resolve/plugins/create/create_shot_clip.py#L120

BigRoy replied:
No - the context data is the "global context data" of the Create context and does not store "per instance". This would still need implementation exactly here on the Host addon with the `IPublishHost` interface

See: https://ayon.ynput.io/docs/dev_publishing#required-functions-in-host-implementation

In Maya we store that in the global file info for example, see [here](https://github.com/ynput/ayon-maya/blob/ca40abe992efce70d9ef26f049a5500513ba0ea7/client/ayon_maya/api/pipeline.py#L158-L170).

robin-ynput replied:
OK, I'll report the TODOs then.

BigRoy commented on client/ayon_resolve/api/plugin.py:
```diff
@@ -404,9 +135,15 @@ def load(self, files):
         self.active_bin = lib.create_bin(self.data["binPath"])
 
         # create mediaItem in active project bin
+
```
```suggestion
```

BigRoy commented on client/ayon_resolve/plugins/publish/collect_plates.py:
```diff
@@ -0,0 +1,27 @@
+        instance.data.update(
+            edit_shared_data[parent_instance_id]
+        )
+
+        self.log.debug(pprint.pformat(instance.data))
```
Can we avoid this heavy verbose logging here?
```suggestion
```

BigRoy commented on client/ayon_resolve/plugins/publish/collect_shots.py:
```diff
@@ -0,0 +1,112 @@
+        )
+
+        self._inject_editorial_shared_data(instance)
+        self.log.debug(pprint.pformat(instance.data))
```
Can we avoid this heavy verbose logging here? Or be more specific about what we want to log?
```suggestion
```

robin-ynput replied:
I'm gonna let @jakubjezek001 answer this one. I've basically reported it from existing plugins for consistency.

From my experience, it's bruteforce but very valuable for a developper (hence debug) especially for troubleshooting.
But if it ends up in user face that'll be quite disturbing for an artist indeed.

BigRoy replied:
If this is the kind of debugging we need - I'd argue that we should instead focus on allowing such functionality inside the Publisher UI itself (similar to the Pyblish Debug Stepper https://github.com/ynput/ayon-core/pull/753) so that one could e.g. have the publisher itself set e.g. debug flags for certain plug-ins and then e.g. add a "profiled" report that e.g. reports the "changes" to the `instance.data` from that plug-in, etc.

We should really try and refrain from having each plug-in enforcing this massive data point all the time. It's slow - and on my old machine; it's **very slow**. The majority of my publish time is spent around these calls ;) (luckily these are without the `json.dumps` around it because those were even slower)

robin-ynput replied:
That makes sense, especially if it impacts performances. I'll remove those, I guess a developer can easily bring them back if needed while troubleshooting.

BigRoy reviewed (approved):
This works! ✅ I did have some remarks on code and one lacking (non-urgent) feature. But other than that, amazing effort. Already so much better than the old creator stuff.

jakubjezek001 commented:
Great progress, @robin-ynput! My testing went well with the 19.5 version. I no longer have 18.5 installed, so this should be enough for now.

We're another step closer to the final state. Here are some points we found during testing:

## Future context with `has_promised_context`
![image](https://github.com/user-attachments/assets/9cfd74b3-41d3-4a51-a2b2-4390481ab344)

In the image, you can see that we are still using the original implementation of editorial. The limits did not allow us to add future context into the instance context. After introducing `has_promised_content`, this no longer applies and can be refactored. The future context should be in the appropriate section, and the instance attribute **Folder path** can be removed. This applies to all highlighted instances. Only the _WorkfileConform_ instance has the correct context.

Since the `has_promised_context` attribute is in place, the Task can be left empty, and the Folder path can be set to a nonexistent value.

iLLiCiTiT commented on client/ayon_resolve/plugins/create/create_workfile.py:
```diff
@@ -0,0 +1,118 @@
+    def _create_new_instance(self):
+        """Create new instance."""
+        variant = self.default_variant
+        project_name = self.create_context.get_current_project_name()
+        folder_path = self.create_context.get_current_folder_path()
+        task_name = self.create_context.get_current_task_name()
+        host_name = self.create_context.host_name
+
+        folder_entity = ayon_api.get_folder_by_path(
+            project_name, folder_path)
+        task_entity = ayon_api.get_task_by_name(
+            project_name, folder_entity["id"], task_name
+        )
```
```suggestion
        project_name = self.create_context.get_current_project_name()
        host_name = self.create_context.host_name
        folder_entity = self.get_current_folder_entity()
        task_entity = self.get_current_task_entity()
        folder_path = folder_entity["path"]
        task_name = task_entity["name"]
        product_name = self.get_product_name(
            project_name,
            folder_entity,
            task_entity,
            self.default_variant,
            host_name,
        )
```

iLLiCiTiT commented on client/ayon_resolve/api/constants.py:
```diff
@@ -0,0 +1,25 @@
+
+# Ayon marker workflow variables
+LEGACY_OPENPYPE_MARKER_NAME = "OpenPypeData"
+AYON_MARKER_NAME = "AyonData"
```
```suggestion
AYON_MARKER_NAME = "AYONData"
```

iLLiCiTiT commented on client/ayon_resolve/api/constants.py:
```diff
@@ -0,0 +1,25 @@
+TEMP_MARKER_FRAME = None
+
+# Ayon default timeline
+AYON_TIMELINE_NAME = "AyonTimeline"
```
```suggestion
AYON_TIMELINE_NAME = "AYONTimeline"
```

iLLiCiTiT commented on client/ayon_resolve/utility_scripts/ayon_startup.scriptlib:
```diff
@@ -1,4 +1,4 @@
--- Run OpenPype's Python launch script for resolve
+-- Run Ayon's Python launch script for resolve
```
```suggestion
-- Run AYON's Python launch script for resolve
```

iLLiCiTiT commented on client/ayon_resolve/utility_scripts/ayon_startup.scriptlib:
```diff
@@ -12,7 +12,7 @@ if ayon_startup_script ~= nil then
         -- We must use RunScript to ensure it runs in a separate
         -- process to Resolve itself to avoid a deadlock for
-        -- certain imports of OpenPype libraries or Qt
+        -- certain imports of Ayon libraries or Qt
```
```suggestion
        -- certain imports of AYON libraries or Qt
```

robin-ynput commented on client/ayon_resolve/plugins/create/create_shot_clip.py:
```diff
@@ -1,245 +1,490 @@
+            EnumDef(
+                "clip_variant",
+                label="Product Variant",
+                tooltip="Chose variant which will be then used for "
```
thanks for that

robin-ynput commented:
@jakubjezek001 I've addressed your feedback regarding the `has_promised_context`. Thanks for that.
![adjusted_feedback_resolve](https://github.com/user-attachments/assets/38551ac3-fe71-45a4-8126-70b425c8556d)

Regarding your question about handles workflow.
So, I have never been an editor myself and always was on the tooling side, but I don't recall any dynamic/min handles workflow. 
As far as I can remember, Editors were creating shots in Ftrack/Shotgrid with handles defined from client specs document and keep it consistent per shot. 
So the logic we have in `extract_otio_review` (extra from media if available or generate black frames otherwise) is what I am more used to.

jakubjezek001 reviewed (approved):
All works fine as expected! Great work @robin-ynput - we can merge this but lets wait also for the merge of https://github.com/ynput/ayon-core/pull/916 and ideally after it is released, so we could change the dependent core version in `ayon-resolve\package.py` under **ayon_required_addons**.

robin-ynput commented:
TODO:
- [ ] Deploy and release https://github.com/ynput/ayon-core/pull/897
- [ ] Adjust `package.py` requirement accordingly

jakubjezek001 commented:
It looks like the linked issue might have been added by mistake. Is it correct? After a quick look at the changes, it seems that the suggested changes from issue #2 weren't applied.

robin-ynput commented:
> It looks like the linked issue might have been added by mistake. Is it correct? After a quick look at the changes, it seems that the suggested changes from issue #2 weren't applied.

This is not a mistake, I've tested all of the work on Windows with OTIO 0.17.0 and 0.16.0 it works. 
So the issue described on this issue has been fixed but with another approach than the suggested change on the ticket (wrapper function + enforce documentation so people use 0.15.0 version).
